Code-the-Dream-School/hh-team4-back
# Error handling fixed

1) What does this PR do? 2) List out what you did.

Created error handling

middleware>

errorHandler.js
imports status code library ( Maps Http status codes to names , example 400 = BAD_REQUEST )
import all custom error classes (for specific errors)
console logs err for Debugging
sets a default status code and message for error response
if logic checks errors name/property to customize a response
Sends a JSON response with message and statusCode
exports the module for errorHandler to be reused in other parts of the app
errors>
+badRequest.js
defines a BadRequestError class that sets a custom message and sets a 400 status code (BAD_REQUEST)
Used for client errors (missing or invalid input)

+createError.js
defines a CreateError class that sets a custom message and sets a 409 status code (CONFLICT)
Used for conflicts , for example trying to create a resource that already exist (duplicate entries)

+customError.js
Creates a custom error class that extends the built-in Error class. To throw errors with custom messages
Used to create a specific error type with a custom message.

+forbidden.js
defines a custom error class ForbiddenError with a default message of "User Access Forbidden" and a status code of 403 (FORBIDDEN)
Used when a user is authenticated but does not have necessary permissions to access a resource (for example trying to access a restricted area)

+notFound.js
defines a custom error class NotFoundError. Extends the CustomError and sets the status code to 404 (NOT_FOUND)
Used for situations where a requested resource is not found.

+unauthenticated.js
defines a custom error class UnauthenticatedError. Extends CustomError Class and sets status to 401 (UNAUTHORIZED)
Used for when a user tries to access a resource without proper authentication (missing or invalid login credentials)

+index.js
Bundles all error classes in one place, then exports all custom error classes.
For easy import and use in other files

Related Issue
Type of Changes
Creation

Acceptance Criteria
Update Screenshots
Provide screenshots of the view before and after the changes.

Before
None

After
Screenshot 2024-11-21 102755

Testing Instructions
testRoutes.js includes routes/scenarios for situations that trigger specific errors. To test enter the route path in the address bar after (see below)
http://localhost:8000/api/v1/test/ for example
test-validation-error
http://localhost:8000/api/v1/test/test-validation-error

Learnings (Optional)
